src/cls.ts

```diff
@@ -43,78 +49,110 @@ export interface PhantomRootContext {
+    } else {
+      if (config.mechanism !== 'async-listener') {
+        if (config.mechanism === 'async-hooks') {
+          this.logger.warn(
+              'TraceCLS#constructor: [async-hooks]-based context',
+              `propagation is not available in Node ${process.version}.`,
+              'Falling back to using async-listener.');
+        } else {
+          this.logger.warn(
+              'TraceCLS#constructor: The specified CLS mechanism',
+              `[${config.mechanism}] was not recognized.`,
+              'Falling back to using async-listener.');
+        }
+      }
+      this.spareCLS = new AsyncListenerCLS(uncorrelated);
+      this.rootSpanStackOffset = 8;
+      this.logger.info(
+          'TraceCLS#constructor: Created [async-listener] CLS instance.');
+    }
+    this.currentCLS = new UniversalCLS(untraced);
+    this.currentCLS.enable();
+  }
 
-export function createNamespace(): CLS.Namespace {
-  return cls.createNamespace(TRACE_NAMESPACE);
-}
+  isEnabled(): boolean {
+    return this.enabled;
+  }
 
-export function destroyNamespace(): void {
-  cls.destroyNamespace(TRACE_NAMESPACE);
-}
+  private rotate(): void {
+    this.currentCLS.disable();
+    const temp = this.currentCLS;
+    this.currentCLS = this.spareCLS;
+    this.spareCLS = temp;
+    this.currentCLS.enable();
+  }
 
-export function getNamespace(): CLS.Namespace {
-  return cls.getNamespace(TRACE_NAMESPACE);
-}
+  enable(): void {
+    if (!this.enabled) {
+      this.logger.info('TraceCLS#enable: Enabling CLS.');
+      this.enabled = true;
+      this.rotate();
+    }
+  }
 
-/**
- * Get a RootContext object from continuation-local storage.
- */
-export function getRootContext(): RootContext {
-  // First getNamespace check is necessary in case any
-  // patched closures escaped before the agent was stopped and the
-  // namespace was destroyed.
-  const namespace = getNamespace();
-  if (namespace) {
-    // A few things can be going on here:
-    // 1. setRootContext has been called earlier to store a real root span
-    //    in continuation-local storage, so retrieve it.
-    // 2. setRootContext has been called earlier to explicitly specify that
-    //    the request corresponding to this continuation is _not_ being traced
-    //    (by being passed UNTRACED_SPAN), so retrieve it as well.
-    // 3. setRootContext has _never_ been called in this continuation. This
-    //    indicates that context was lost, and namespace.get('root') will
-    //    return null. Therefore, explicitly return UNCORRELATED_SPAN to
-    //    indicate that context was lost.
-    return namespace.get('root') || UNCORRELATED_SPAN;
-  } else {
-    // No namespace indicates that the Trace Agent is disabled. This is a
-    // special case where _all_ requests are explicitly not being traced,
-    // so return UNTRACED_SPAN to be consistent with that.
-    return UNTRACED_SPAN;
+  disable(): void {
```

Comment:
I don't like the disable mechanism too much. My expected outcome is that at the end of `disable`, we have set `this.currentCLS` to `UnversalCLS`. That's what the code does, but it is not dead obvious. It should be dead obvious.

Instead of the `{current,spare}CLS` mechanism which points to long-lived instances, can you instead have a property that points the the `class`? `enable` should create an instance of that class, and `disable` should destroy. `isEnabled` can be a test on the non-nullness of `this.cls` instead of a boolen. Thoughts?

Reply:
Agree, what do you think of the change?